File: app/src/test/java/org/gnucash/android/test/unit/BookHelperTest.java
package org.gnucash.android.test.unit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

import android.database.sqlite.SQLiteDatabase;

import org.gnucash.android.app.GnuCashApplication;
import org.gnucash.android.db.DatabaseHelper;
import org.gnucash.android.db.adapter.AccountsDbAdapter;
import org.gnucash.android.db.adapter.BooksDbAdapter;
import org.gnucash.android.db.adapter.RecurrenceDbAdapter;
import org.gnucash.android.db.adapter.ScheduledActionDbAdapter;
import org.gnucash.android.db.adapter.SplitsDbAdapter;
import org.gnucash.android.db.adapter.TransactionsDbAdapter;
import org.gnucash.android.importer.GncXmlHandler;
import org.junit.After;
import org.junit.Before;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.XMLReader;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;

import timber.log.Timber;

public abstract class BookHelperTest {
    protected SQLiteDatabase mImportedDb;
    protected BooksDbAdapter mBooksDbAdapter;
    protected TransactionsDbAdapter mTransactionsDbAdapter;
    protected AccountsDbAdapter mAccountsDbAdapter;
    protected ScheduledActionDbAdapter mScheduledActionDbAdapter;

    protected String importGnuCashXml(String filename) {
        SAXParser parser;
        GncXmlHandler handler = null;
        try {
            parser = SAXParserFactory.newInstance().newSAXParser();
            XMLReader reader = parser.getXMLReader();
            handler = new GncXmlHandler();
            reader.setContentHandler(handler);
            InputStream inputStream = getClass().getClassLoader().getResourceAsStream(filename);
            InputSource inputSource = new InputSource(new BufferedInputStream(inputStream));
            reader.parse(inputSource);
        } catch (ParserConfigurationException | SAXException | IOException e) {
            Timber.e(e);
            fail();
        }
        String bookUID = handler.getBookUID();
        setUpDbAdapters(bookUID);
        return bookUID;
    }

    private void setUpDbAdapters(String bookUID) {
        DatabaseHelper databaseHelper = new DatabaseHelper(GnuCashApplication.getAppContext(), bookUID);
        SQLiteDatabase mainDb = databaseHelper.getReadableDatabase();
        mTransactionsDbAdapter = new TransactionsDbAdapter(mainDb, new SplitsDbAdapter(mainDb));
        mAccountsDbAdapter = new AccountsDbAdapter(mainDb, mTransactionsDbAdapter);
        RecurrenceDbAdapter recurrenceDbAdapter = new RecurrenceDbAdapter(mainDb);
        mScheduledActionDbAdapter = new ScheduledActionDbAdapter(mainDb, recurrenceDbAdapter);
        mImportedDb = mainDb;
    }

    @Before
    public void setUp() throws Exception {
        mBooksDbAdapter = BooksDbAdapter.getInstance();
        mBooksDbAdapter.deleteAllRecords();
        assertThat(mBooksDbAdapter.getRecordsCount()).isZero();
    }

    @After
    public void tearDown() throws Exception {
        if (mTransactionsDbAdapter != null) {
            mTransactionsDbAdapter.close();
            mTransactionsDbAdapter = null;
        }
        if (mAccountsDbAdapter != null) {
            mAccountsDbAdapter.close();
            mAccountsDbAdapter = null;
        }
        if (mScheduledActionDbAdapter != null) {
            mScheduledActionDbAdapter.close();
            mScheduledActionDbAdapter = null;
        }
        if (mImportedDb != null) {
            mImportedDb.close();
            mImportedDb = null;
        }
    }
}

File: app/src/test/java/org/gnucash/android/test/unit/BookHelperTest.kt
package org.gnucash.android.test.unit

import org.assertj.core.api.Assertions.assertThat
import org.gnucash.android.BuildConfig
import org.gnucash.android.db.DatabaseHelper
import org.gnucash.android.db.DatabaseHolder
import org.gnucash.android.db.adapter.AccountsDbAdapter
import org.gnucash.android.db.adapter.BooksDbAdapter
import org.gnucash.android.db.adapter.BudgetsDbAdapter
import org.gnucash.android.db.adapter.CommoditiesDbAdapter
import org.gnucash.android.db.adapter.PricesDbAdapter
import org.gnucash.android.db.adapter.RecurrenceDbAdapter
import org.gnucash.android.db.adapter.ScheduledActionDbAdapter
import org.gnucash.android.db.adapter.TransactionsDbAdapter
import org.gnucash.android.importer.GncXmlImporter
import org.gnucash.android.util.ConsoleTree
import org.junit.After
import org.junit.Before
import timber.log.Timber
import java.nio.charset.StandardCharsets

abstract class BookHelperTest : GnuCashTest() {
    protected lateinit var dbHolder: DatabaseHolder
    protected lateinit var booksDbAdapter: BooksDbAdapter

    protected lateinit var transactionsDbAdapter: TransactionsDbAdapter
    protected lateinit var accountsDbAdapter: AccountsDbAdapter
    protected lateinit var scheduledActionDbAdapter: ScheduledActionDbAdapter
    protected lateinit var commoditiesDbAdapter: CommoditiesDbAdapter
    protected lateinit var budgetsDbAdapter: BudgetsDbAdapter
    protected lateinit var pricesDbAdapter: PricesDbAdapter

    protected fun importGnuCashXml(filename: String): String {
        val inputStream = openResourceStream(filename)
        val bookUID = GncXmlImporter.parse(context, inputStream)
        setUpDbAdapters(bookUID)
        return bookUID
    }

    private fun setUpDbAdapters(bookUID: String) {
        close()
        val databaseHelper = DatabaseHelper(context, bookUID)
        val mainHolder = databaseHelper.holder
        commoditiesDbAdapter = CommoditiesDbAdapter(mainHolder)
        transactionsDbAdapter = TransactionsDbAdapter(commoditiesDbAdapter)
        accountsDbAdapter = AccountsDbAdapter(transactionsDbAdapter)
        val recurrenceDbAdapter = RecurrenceDbAdapter(mainHolder)
        scheduledActionDbAdapter =
            ScheduledActionDbAdapter(recurrenceDbAdapter, transactionsDbAdapter)
        budgetsDbAdapter = BudgetsDbAdapter(recurrenceDbAdapter)
        pricesDbAdapter = PricesDbAdapter(commoditiesDbAdapter)
        dbHolder = mainHolder
    }

    @Before
    open fun setUp() {
        System.gc()
        booksDbAdapter = BooksDbAdapter.instance
        assertThat(booksDbAdapter.recordsCount).isOne()
        setUpDbAdapters(booksDbAdapter.activeBookUID)
    }

    @After
    open fun tearDown() {
        close()
    }

    private fun close() {
        if (::accountsDbAdapter.isInitialized) accountsDbAdapter.close()
        if (::budgetsDbAdapter.isInitialized) budgetsDbAdapter.close()
        if (::commoditiesDbAdapter.isInitialized) commoditiesDbAdapter.close()
        if (::pricesDbAdapter.isInitialized) pricesDbAdapter.close()
        if (::scheduledActionDbAdapter.isInitialized) scheduledActionDbAdapter.close()
        if (::transactionsDbAdapter.isInitialized) transactionsDbAdapter.close()
        if (::dbHolder.isInitialized) dbHolder.close()
    }

    protected fun readFile(name: String): String {
        val stream = openResourceStream(name)
        val bytes = stream.readAllBytes()
        return String(bytes, StandardCharsets.UTF_8)
    }

    companion object {
        init {
            Timber.plant(ConsoleTree(BuildConfig.DEBUG) as Timber.Tree)
        }
    }
}
